tensorflow/python/framework/python_op_gen.cc

```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
+      param_type_map[attr.name()] = type_var_name;
+    }
+  }
+
+  for (int i = 0; i < api_def_.arg_order_size(); ++i) {
+    const auto& arg = *FindInputArg(api_def_.arg_order(i), op_def_);
+    // Do not add type annotations to args that accept multiple tensors
+    if (!arg.number_attr().empty()) continue;
+    string type_annotation;
+    if (param_type_map.find(arg.type_attr()) != param_type_map.end()) {
+      // Get the correct TypeVar if input maps to an attr
+      strings::StrAppend(&type_annotation, "_ops.Tensor[", param_type_map[arg.type_attr()], "]");
+    } else {
+      // Get the dtype of the tensor
+      const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+      if (dtypes_map.find(py_dtype) != dtypes_map.end()) {
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtypes_map[py_dtype], "]");
+      }
+    }
+
+    param_type_map[arg.name()] = type_annotation;
+  }
+
+  // Add type annotations to params
+  string parameters;
+  for (const auto& param : params_no_default_) {
+    if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
+    strings::StrAppend(&parameters, param.GetRenameTo());
+
+    if (param_type_map.find(param.GetName()) != param_type_map.end()) {
+      if(!param_type_map[param.GetName()].empty()){
+        strings::StrAppend(&parameters, ": ", param_type_map[param.GetName()]);
+      }
+    }
+  }
+
+  return parameters;
+}
+
+// Generate TypeVars using attrs
+void GenEagerPythonOp::GenerateTypeVars() {
+  bool added_typevar = false;
+  for (int i = 0; i<op_def_.attr_size(); ++i) {
+    const auto& attr(op_def_.attr(i));
+    if (attr.type() == "type") {
+      std::vector<string> allowed_types;
+      bool has_dtype_half = false;
+      for (int t : attr.allowed_values().list().type()) {
+        if (t == 19) { // DT_HALF = 19
+          has_dtype_half = true;
+          break;
+        }
+        DataType dtype = static_cast<DataType>(t);
+        const string py_dtype = python_op_gen_internal::DataTypeToPython(dtype, "_dtypes.");
+        if (dtypes_map.find(py_dtype) != dtypes_map.end()) {
+          allowed_types.emplace_back(dtypes_map[py_dtype]);
+        }
+      }
+
+      // Do not create a type variable that includes the dtype half
+      if (has_dtype_half) continue;
+
+      // If all dtypes are allowed, add them all
+      if (allowed_types.empty()) {
+        for (std::pair<string, string> map_dtype : dtypes_map) {
+          allowed_types.emplace_back(map_dtype.second);
+        }
+      }
+
+      std::sort(allowed_types.begin(), allowed_types.end());
+
+      string typevar_dtypes;
+      for (std::vector<string>::iterator it = allowed_types.begin(); it != allowed_types.end(); ++it) {
+        if (!typevar_dtypes.empty()) strings::StrAppend(&typevar_dtypes, ", ");
+        strings::StrAppend(&typevar_dtypes, *it);
+      }
+
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
```

Comment:
It seems that the type var name is generated twice - let's only generate it once, and keep a map from attribute name to type var name.